File: demos_code_J2_conception_objet/06_break_package_cycle/apres/src/com/acme/billing/BillingService.java
package com.acme.billing;
import com.acme.contracts.RefundListener;

public class BillingService {
    private final RefundListener listener;
    public BillingService(RefundListener listener) { this.listener = listener; }
    public void refund(String id) {
        System.out.println("billing: refund " + id);
        listener.refunded(id);
    }
}

File: Demo en Python/demos_code_J2_conception_objet_python/06_break_package_cycle/apres/src/com/acme/billing/billing_service.py
from com.acme.contracts.refund_listener import RefundListener


class BillingService:
    def __init__(self, listener: RefundListener) -> None:
        self._listener = listener

    def refund(self, order_id: str) -> None:
        print(f"billing: refund {order_id}")
        self._listener.refunded(order_id)
